### ISSUE TITLE
Play/Stop/Forward/Backward function from AirPlay Speakers

### ISSUE THREAD

**Issue by Ogekuri** [ejurgensen/forked-daapd#531]
Hi,
I've try to use **Play**/**Stop**/**Forward**/**Backward** functions from my AirPlay Speakers (a DENON AVR-X2300W) but it doesn't work.

There is something to configure or isn't already supported?

Many Thanks!

_-- 
Oge_

**Comment by ejurgensen** [comment]
Your device is probably not sending the commands to forked-daapd, like it does to iTunes. I have the same with my Sony. Why it is not doing so is not clear to me. It could be some flag that needs to be set in an announcement, or it could be some auth that needs to be set up before the device will issue the commands.

I will try to investigate, but it might be difficult to solve.

**Comment by Ogekuri** [comment]
Ok, I know, no problem.

Here I've captured the packets of **Forward** and **Backward** commands but they seem sent to the avahi-daemon on 5353. Make sense for you?

**Forward**

`2018-05-14 11:33:36.129892 IP 192.168.1.115.5353 > 224.0.0.251.5353: 0 [2q] SRV (QM)? iTunes_Ctrl_B5BDB5DF964C8D67._dacp._tcp.local. TXT (QM)? iTunes_Ctrl_B5BDB5DF964C8D67._dacp._tcp.local. (69)
        0x0000:  4500 0061 4438 0000 ff11 d43c c0a8 0173  E..aD8.....<...s
        0x0010:  e000 00fb 14e9 14e9 004d d74b 0000 0000  .........M.K....
        0x0020:  0002 0000 0000 0000 1c69 5475 6e65 735f  .........iTunes_
        0x0030:  4374 726c 5f42 3542 4442 3544 4639 3634  Ctrl_B5BDB5DF964
        0x0040:  4338 4436 3705 5f64 6163 7004 5f74 6370  C8D67._dacp._tcp
        0x0050:  056c 6f63 616c 0000 2100 01c0 0c00 1000  .local..!.......
        0x0060:  01                                       .
2018-05-14 11:33:38.129797 IP 192.168.1.115.5353 > 224.0.0.251.5353: 0 [2q] SRV (QM)? iTunes_Ctrl_B5BDB5DF964C8D67._dacp._tcp.local. TXT (QM)? iTunes_Ctrl_B5BDB5DF964C8D67._dacp._tcp.local. (69)
        0x0000:  4500 0061 443d 0000 ff11 d437 c0a8 0173  E..aD=.....7...s
        0x0010:  e000 00fb 14e9 14e9 004d d74b 0000 0000  .........M.K....
        0x0020:  0002 0000 0000 0000 1c69 5475 6e65 735f  .........iTunes_
        0x0030:  4374 726c 5f42 3542 4442 3544 4639 3634  Ctrl_B5BDB5DF964
        0x0040:  4338 4436 3705 5f64 6163 7004 5f74 6370  C8D67._dacp._tcp
        0x0050:  056c 6f63 616c 0000 2100 01c0 0c00 1000  .local..!.......
        0x0060:  01  `

**Backward**

`2018-05-14 11:34:11.132033 IP 192.168.1.115.5353 > 224.0.0.251.5353: 0 [2q] SRV (QM)? iTunes_Ctrl_B5BDB5DF964C8D67._dacp._tcp.local. TXT (QM)? iTunes_Ctrl_B5BDB5DF964C8D67._dacp._tcp.local. (69)
        0x0000:  4500 0061 448b 0000 ff11 d3e9 c0a8 0173  E..aD..........s
        0x0010:  e000 00fb 14e9 14e9 004d d74b 0000 0000  .........M.K....
        0x0020:  0002 0000 0000 0000 1c69 5475 6e65 735f  .........iTunes_
        0x0030:  4374 726c 5f42 3542 4442 3544 4639 3634  Ctrl_B5BDB5DF964
        0x0040:  4338 4436 3705 5f64 6163 7004 5f74 6370  C8D67._dacp._tcp
        0x0050:  056c 6f63 616c 0000 2100 01c0 0c00 1000  .local..!.......
        0x0060:  01                                       .
2018-05-14 11:34:13.131947 IP 192.168.1.115.5353 > 224.0.0.251.5353: 0 [2q] SRV (QM)? iTunes_Ctrl_B5BDB5DF964C8D67._dacp._tcp.local. TXT (QM)? iTunes_Ctrl_B5BDB5DF964C8D67._dacp._tcp.local. (69)
        0x0000:  4500 0061 4490 0000 ff11 d3e4 c0a8 0173  E..aD..........s
        0x0010:  e000 00fb 14e9 14e9 004d d74b 0000 0000  .........M.K....
        0x0020:  0002 0000 0000 0000 1c69 5475 6e65 735f  .........iTunes_
        0x0030:  4374 726c 5f42 3542 4442 3544 4639 3634  Ctrl_B5BDB5DF964
        0x0040:  4338 4436 3705 5f64 6163 7004 5f74 6370  C8D67._dacp._tcp
        0x0050:  056c 6f63 616c 0000 2100 01c0 0c00 1000  .local..!.......
        0x0060:  01 `

**Comment by Ogekuri** [comment]
> Yes!, tomorrow I'll try this version. Thanks!

Here the results of my test with a **Denon AVR-X2300W**:
Play/Stop: **working**
Forward: **working**
Backward: **working**

Please, let me know here when this branch will be merged so I'll rebuild from the master.

Many thanks for you support.

_-- 
Ciao!
Oge_

This is the version tested from abobe link:
`Forked Media Server: Version 26.0
Copyright (C) 2009-2015 Julien BLACHE <jb@jblache.org>
Based on mt-daapd, Copyright (C) 2003-2007 Ron Pedde <ron@pedde.com>
Released under the GNU General Public License version 2 or later
`

**Comment by LordMyschkin** [comment]
it is a pity it does not work with my Samsung E750  speaker - any way I can debug that as an idiot?
Id' also LOVE to use my remote for controling forked-daapd.

**Comment by ejurgensen** [comment]
The Samsung works with iTunes? If so, you can check if it is making any requests to forked-daapd. Set the log level to debug and try change tracks or stop. If it isn't, which is likely, it becomes more difficult...

Note that the speaker has to be on a network that is configured as trusted in forked-daapd.conf. Otherwise the requests will be denied. But if you have everything on a standard 192.168 network then that is probably not the issue.

**Comment by chme** [comment]
I guess issue #202 is related to this. There is a link to an unofficial documentation https://nto.github.io/AirPlay.html#audio-remotecontrol that may help to get the speaker id to control the volume.

**Comment by ejurgensen** [comment]
Yes, I had somewhat forgotten that issue, embarrasing :-)

The issue stalled on speaker id, which is still a challenge. The request does not include it. Active-Remote could be used if you could assign it per speaker, but I reckon that is not possible?

The only option I can think of is getting the IP address and then get it converted to speaker id by the player/raop. Kind of clunky, though. In principle you could have multiple speakers with the same IP.

Can you think of a better way?

**Comment by chme** [comment]
In commit https://github.com/Japanuspus/forked-daapd/commit/ee7f17e1244c92839c99e05dac25bdc535e5998b you described a possible solution. If i understand it correctly it would allow identifying the session and the session should hold the speaker id. But I have to admit that I did not look into it any further than reading the comments.

**Comment by ejurgensen** [comment]
Yes, but it was based on the assumption that the id would be in the header, and I can see now that it isn’t.

I notice iTunes makes the announcement when the speaker is turned on, not when you start iTunes. Maybe a way to give ActiveRemote to a particular device? But it is still an announcement so not sure how that would work. 

Maybe I will just do it with IP address, should work in most cases.

**Comment by chme** [comment]
> Yes, but it was based on the assumption that the id would be in the header, and I can see now that it isn’t.

Do you mean the "Active-Remote" header is missing in the request from the speaker? In current master we do not add the header to the raop headers, that might be the reason the speaker does not add one either. 
In https://github.com/Japanuspus/forked-daapd/commit/ee7f17e1244c92839c99e05dac25bdc535e5998b#diff-ffbc3a37e2ce095ff7af5ad92105f054R1148 the "Active-Remote" header is added in raop_add_headers. If i understand the documentation correctly forked-daapd should include the header in all requests to the speaker and the speaker includes this header in its requests to the control port.

**Comment by chme** [comment]
And maybe a look at shairport-sync might give some further insight. With version 3.2 it gained support to send these kind of commands to the server (MPRIS interface).

**Comment by ejurgensen** [comment]
Boy my memory on this is bad, thanks for reminding me again! I meant speaker id, not Active-Remote, but like you say it is just a matter of sending that to the device with the speaker id value. I have not observed missing Active-Remote values. For some reason I was thinking that the device got the value from the mdns announcement.

**Comment by ejurgensen** [comment]
I have now implemented Active-Remote in the master branch (with commit b6e1269). This means that volume changes from Airplay speakers are received and understood by forked-daapd, and at least dacp clients like Remote will update their speaker volume settings automatically.

It seems Active-Remote is also necessary for proper support of speaker play/stop/forward/back/etc, so my first change wasn't complete. After a while, my Sony stopped working with that, and I found out that the reason it had worked was because it used the Active-Remote value from iTunes (which I had running while looking into this initially). After stopping iTunes, and power-cycling the Sony, it would no longer make requests, probably because it had no Active-Remote value (as the spec @chme linked to also kind of says, so I should probably have read that better).

Now, with Active-Remote from forked-daapd the Sony works without iTunes running and it seems stable. @LordMyschkin I have hopes that this also fixes your Samsung, so perhaps you can try with the latest master?

**Comment by LordMyschkin** [comment]
No, Im sorry it does not, but I honestly believe the controls only control the hardware dock (I cannot test  and never used - long time before lightning port...) 
I did not found a lot of documentation about this speaker (i bought on ebay a while ago)....

**Comment by Ogekuri** [comment]
> After stopping iTunes, and power-cycling the Sony, it would no longer make requests, probably because it had no Active-Remote value (as the spec @chme linked to also kind of says, so I should probably have read that better).

>Now, with Active-Remote from forked-daapd the Sony works without iTunes running and it seems stable

Hi  ejurgensen,

I've seen a same thing on my side.

I've already rebuilt form lastest master and it's seem more stable.

Many thanks!
Ciao!

**Comment by mrhund** [comment]
Hi, 
my HomePod obviously sends the commands made by the hardware controls of the speaker back to forked-daapd, but I have the following in my log:

dacp: No session-id specified in request

Any ideas about that?

Version: main: Forked Media Server Version 26.2 via apt on Raspbian Jessie

**Comment by ejurgensen** [comment]
Check that your HomePod is on a network that is configured as trusted in forked-daapd’s config (see trusted_networks). Otherwise the request gets denied because the device can’t authorize itself.

**Comment by mrhund** [comment]
Thank you for the quick response. 

AppleTV and Homepod are in the same network and so both are in a trusted network. 

But your comment with the network brought me to an idea and so I **diabled IPv6** in den config and now it works.
Well this is one solution, the other would have been adding thw IPv6 addresses also to the trusted networks.

**Comment by LordMyschkin** [comment]
I just noticed with true delight that with the current version of forked-daapd (git from 21.9.2018), my Samsung E750 works with remote, and also my Boston MC200Air.
This is really a HUGE improvement!!! Love it!
